src/mod/asr_tts/mod_unimrcp/mod_unimrcp.c

```diff
@@ -1873,10 +1873,15 @@ static apt_bool_t speech_on_channel_add(mrcp_application_t *application, mrcp_se
 	speech_channel_t *schannel = (speech_channel_t *) mrcp_application_channel_object_get(channel);
 	char codec_name[60] = { 0 };
 	const mpf_codec_descriptor_t *descriptor;
+	switch_channel_t *orig_channel = NULL;
 
 	/* check status */
 	if (!session || !schannel || status != MRCP_SIG_STATUS_CODE_SUCCESS) {
 		goto error;
+	} else {
+		switch_core_session_t *orig_session = switch_core_memory_pool_get_data(schannel->memory_pool, "__session");
+		if (orig_session)
+			orig_channel = switch_core_session_get_channel(orig_session);
```

Comment:
speech_on_channel_add is a event coming from the lib/unimrcp. We are not sure when the event is fired, the session are still alive. I would use the speech_channel_t->session_uuid to get the session and lock it up so we know that session is alive at this point. Then get all the data you need (line 1926 - 1929) to be included in the event then release the session lock.

Reply:
Good point. I'll make the change

Reply:
Done

Reply:
On line 1896, if that code hits then it will cause a session deadlock. Either you copy/dup those variable or move this code inside line 1917

Reply:
Moved